### PR TITLE
fix(ts): export diff type files for cjs and esm

### LINKED ISSUE #355: `package.json` exports and types are not compatible with TypeScript's `node16` `moduleResolution`
### Description

TypeScript added a new module resolution strategy `node16` that better mimicks Node.js. Only under this option (or `nodenext`) is the `exports` field evaluated.

When authoring a `.ts` file in a CommonJS project and using this import
```ts
import rateLimit from 'express-rate-limit'
```
TypeScript complains about trying to require an ES module from CommonJS. This is because there's only one `.d.ts` file for both the CJS and ESM entrypoint and TypeScript considers that to be ESM due to the nearest `package.json`.
https://www.typescriptlang.org/docs/handbook/esm-node.html#packagejson-exports-imports-and-self-referencing mentions at the bottom that each needs their own declaration file.

As awkward as this is, this can be fixed by
1. renaming `index.d.ts` to `index.d.mts`
2. copying `index.d.mts` to `index.d.cts`
3. changing the `package.json` exports like this:
```
	"exports": {
		".": {
			"import": {
				"types": "./dist/index.d.mts",
				"default": "./dist/index.mjs"
			},
			"require": {
				"types": "./dist/index.d.cts",
				"default": "./dist/index.cjs"
			}
		}
	},
```

### Library version

6.7.0

### Node version

16.20.0

### Typescript version (if you are using it)

5.1.6

### Module system

CommonJS

### PR DESCRIPTION
**Pull request by gamemaker1**
## Related Issues

- Fixes #355

### Changed

- Bumps test project dependencies.
- Changes `export` field in package.json to export different `.d.ts` files for CommonJS and ESM projects.

## Checklist

- [x] The issues that this PR fixes/closes have been mentioned above.
- [x] What this PR adds/changes/removes has been explained.
- [x] All tests (`npm test`) pass.
- [x] All added/modified code has been commented, and
      methods/classes/constants/types have been annotated with TSDoc comments.
- [x] If a new feature has been added or a bug has been fixed, tests have been
      added for the same.

### CODE REVIEW

File: test/external/imports/named-import/ts-cjs/source/index.ts [hunk 1]
```diff
@@ -1,7 +1,7 @@
 // /source/index.ts
 // Run the server
 
-import app from './app'
+import { app } from './app'
```
Comment (nfriedly) [inline test/external/imports/named-import/ts-cjs/source/index.ts:4]: how did that work before...?

**Review by nfriedly — APPROVED** [review]
I think we can leave it as-is, but it would have been easier to review this a series of PRs: one for the dependency updates and associated script, one for node16 types fix, and maybe one or two more for the app import statements and autofix tweaks.

**Comment by nfriedly** [comment]
@AlCalzone do you think you can check this branch out and test it locally to confirm that it works for you before we ship it?

It's been a bit since I've done this, but I believe the process is something along the lines of this:

```sh
cd ~
git clone https://github.com/express-rate-limit/express-rate-limit.git
cd express-rate-limit
git checkout fix/diff-dts-files
npm install
npm run compile
npm link
cd <directory where your project is>
npm link express-rate-limit
```

(The `npm link` command lets you use a local version of a dependency instead of the one installed from npm - see https://docs.npmjs.com/cli/v9/commands/npm-link for more.)

Once you're done testing the fix, I believe a regular `npm install express-rate-limit` will undo the link.

**Comment by gamemaker1** [comment]
You are right, I should have done that.

I had to update the dependencies to reproduce the bug first, so I added the `upgrade_dependencies` function to the script and did the autofix thing 😅

File: test/external/imports/named-import/ts-cjs/source/index.ts (hunk 1 above)
Comment (gamemaker1) [reply test/external/imports/named-import/ts-cjs/source/index.ts:4]: I was shocked too, I have literally no idea how it worked before

**Comment by AlCalzone** [comment]
Seems to be working

**Comment by nfriedly** [comment]
Awesome, v6.7.1 is now live on npm with the fix.

**Pull request merged**